test/studies/shootout/submitted/knucleotide4.chpl

```diff
@@ -0,0 +1,140 @@
+/* The Computer Language Benchmarks Game
+   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
+
+   contributed by Ben McDonald
+   derived from the Chapel #3 version by Ben Harshbarger and Brad Chamberlain
+*/
+
+use IO, Map, Sort;
+
+config param columns = 61;
+
+
+proc main(args: [] string) {
+  // Open stdin and a binary reader channel
+  const consoleIn = openfd(0),
+        fileLen = consoleIn.size,
+        stdinNoLock = consoleIn.reader(kind=ionative, locking=false);
+
+  // Read line-by-line until we see a line beginning with '>TH'
+  var buff: [1..columns] uint(8),
+      lineSize = 0,
+      numRead = 0;
+
+  while stdinNoLock.readline(buff, lineSize) && !startsWithThree(buff) do
+    numRead += lineSize;
+
+  // Read in the rest of the file
+  var dataDom = {1..fileLen-numRead},
+      data: [dataDom] uint(8),
+      idx = 1;
+
+  while stdinNoLock.readline(data, lineSize, idx) do
+    idx += lineSize - 1;
+
+  // Resize our array to the amount actually read
+  dataDom = {1..idx};
+
+  // Make everything uppercase
+  forall d in data do
+    d -= ("a".toByte() - "A".toByte());
+
+  writeFreqs(data, 1);
+  writeFreqs(data, 2);
+  writeCount(data, "GGT");
+  writeCount(data, "GGTA");
+  writeCount(data, "GGTATT");
+  writeCount(data, "GGTATTTTAATT");
+  writeCount(data, "GGTATTTTAATTTATAGT");
+}
+
+
+proc writeFreqs(data, param nclSize) {
+  const freqs = calculate(data, nclSize);
+
+  // create an array of (frequency, sequence) tuples
+  var arr = for (s,f) in freqs.items() do (f,s);
+
+  // print the array, sorted by decreasing frequency
+  for (f, s) in arr.sorted(reverseComparator) do
+   writef("%s %.3dr\n", decode(s, nclSize),
+           (100.0 * f) / (data.size - nclSize));
```

Comment:
For this call to decode, I might just unwrap the hash value at the point of creating the array:

```chapel
  var arr = for (s,f) in freqs.items() do (f,s);
```

Alternatively, we could unwrap it as we pass `s` into `decode()` (but I like the first way better if it works).

Without doing the first approach, the thing that's giving me pause is "How does `sorted()` know how to sort `hashVals()`?  Does the compiler come up with a default comparator for them?  (and if so, how does it know that it's right?  And do we want to rely on it?).  By changing `arr` back into raw `int`s we avoid that whole question.

Reply:
Since the hashVal element is second in the tuple here, I don't believe that it is used in any comparison for the sort at all, since sort on a list of tuples chooses the first element to use as a comparator by default. I still like the idea of unwrapping at the array creation though and will switch to that.